docker/Dockerfile

```diff
@@ -17,51 +24,122 @@ RUN apt-get update && apt-get install -y libgeos-dev \
+
+##
+# Create layer with farmOS dev sources.
+FROM farmos-sources as farmos-dev-sources
+
+# Install sources.
+RUN composer update
+
+##
+# Dependencies layer.
+FROM baseimage as farmos-baseimage
 
 # Set Apache ServerName directive globally to suppress AH00558 message.
 RUN echo "ServerName localhost" >> /etc/apache2/apache2.conf
 
-# Install and enable geos.
+# Enable PHP dependencies.
 COPY --from=php-dependencies /usr/local/lib/php/extensions/ /usr/local/lib/php/extensions/
 RUN docker-php-ext-enable geos bcmath
 
 # Add custom PHP configurations.
 COPY conf.d/ /usr/local/etc/php/conf.d
 
-# Install apt dependencies.
-RUN apt-get update && apt-get install -y \
-    # Install git and unzip (needed by Composer).
-    git unzip \
+# Install apt dependencies and clean up.
+RUN apt-get update && apt-get install -y --no-install-recommends\
     # Install postgresql-client so Drush can connect to the database.
     postgresql-client \
     # Install libgeos-c1v5 so geos php extension can use libgeos_c.so.1.
     libgeos-c1v5 \
-  && rm -rf /var/lib/apt/lists/* \
-  && apt-get clean
-
-# Set the COMPOSER_MEMORY_LIMIT environment variable to unlimited.
-ENV COMPOSER_MEMORY_LIMIT=-1
-
-# Add the build-farmOS.sh script.
-COPY build-farmOS.sh /usr/local/bin/
-RUN chmod a+x /usr/local/bin/build-farmOS.sh
-
-# Build the farmOS codebase in /var/farmoS with the --no-dev flag.
-# Change the ownership of the sites directory and copy the farmOS codebase into /opt/drupal.
-RUN mkdir /var/farmOS \
-  && /usr/local/bin/build-farmOS.sh --no-dev \
-  && chown -R www-data:www-data /var/farmOS/web/sites \
-  && rm -r /opt/drupal && cp -rp /var/farmOS /opt/drupal
+    # Install git and unzip (needed by Composer).
+    git unzip \
+    && rm -rf /var/lib/apt/lists/* \
+    && apt-get clean \
+    # Clean up ${DRUPAL_PATH}.
+    && rm -r ${DRUPAL_PATH} \
+    && mkdir ${DRUPAL_PATH}
 
 # Set the entrypoint.
-COPY docker-entrypoint.sh /usr/local/bin/
+COPY --chown=www-data docker-entrypoint.sh /usr/local/bin/
+
+# Set the working directory, entrypoint and command.
+WORKDIR ${DRUPAL_PATH}
 ENTRYPOINT ["docker-entrypoint.sh"]
 CMD ["apache2-foreground"]
+
+##
+# Development image.
+FROM farmos-baseimage as dev
+
+# Change the user/group IDs of www-data inside the image to match the ID of the
+# developer's user on the host machine. This allows Composer to create files
+# owned by www-data inside the container, while keeping those files editable by
+# the developer outside of the container.
+# This defaults to 1000, based on the assumption that the developer is running
+# as UID 1000 on the host machine. It can be overridden at image build time with:
+# --build-arg WWW_DATA_ID=$(id -u)
+ARG WWW_DATA_ID=1000
+RUN usermod -u ${WWW_DATA_ID} www-data && groupmod -g ${WWW_DATA_ID} www-data
+
+# Add farmOS sources
+COPY --from=composer-file --chown=www-data:www-data ${FARMOS_PATH} ${FARMOS_PATH}
```

Comment:
This line can be removed, few lines below dev sources are added that already include the composer.json file generated in composer-file stage.